samples/subsys/usb/uac2_implicit_feedback/src/main.c

```diff
@@ -0,0 +1,570 @@
+/*
+ * Copyright (c) 2023-2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdint.h>
+#include <stdlib.h>
+
+#include <sample_usbd.h>
+#include "feedback.h"
+
+#include <zephyr/cache.h>
+#include <zephyr/device.h>
+#include <zephyr/usb/usbd.h>
+#include <zephyr/usb/class/usbd_uac2.h>
+#include <zephyr/drivers/i2s.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(uac2_sample, LOG_LEVEL_INF);
+
+#define HEADPHONES_OUT_TERMINAL_ID UAC2_ENTITY_ID(DT_NODELABEL(out_terminal))
+#define MICROPHONE_IN_TERMINAL_ID UAC2_ENTITY_ID(DT_NODELABEL(in_terminal))
+
+#define SAMPLES_PER_SOF     48
+#define SAMPLE_FREQUENCY    (SAMPLES_PER_SOF * 1000)
+#define SAMPLE_BIT_WIDTH    16
+#define NUMBER_OF_CHANNELS  2
+#define BYTES_PER_SAMPLE    DIV_ROUND_UP(SAMPLE_BIT_WIDTH, 8)
+#define BYTES_PER_SLOT      (BYTES_PER_SAMPLE * NUMBER_OF_CHANNELS)
+#define MIN_BLOCK_SIZE      ((SAMPLES_PER_SOF - 1) * BYTES_PER_SLOT)
+#define BLOCK_SIZE          (SAMPLES_PER_SOF * BYTES_PER_SLOT)
+#define MAX_BLOCK_SIZE      ((SAMPLES_PER_SOF + 1) * BYTES_PER_SLOT)
+
+/* Absolute minimum is 5 TX buffers (1 actively consumed by I2S, 2nd queued as
+ * next buffer, 3rd acquired by USB stack to receive data to, and 2 to handle
+ * SOF/I2S offset errors), but add 2 additional buffers to prevent out of memory
+ * errors when USB host decides to perform rapid terminal enable/disable cycles.
+ */
+#define I2S_BLOCKS          7
+K_MEM_SLAB_DEFINE_STATIC(i2s_tx_slab, MAX_BLOCK_SIZE, I2S_BLOCKS, 4);
+K_MEM_SLAB_DEFINE_STATIC(i2s_rx_slab, MAX_BLOCK_SIZE, I2S_BLOCKS, 4);
+
+struct usb_i2s_ctx {
+	const struct device *i2s_dev;
+	bool headphones_enabled;
+	bool microphone_enabled;
+	bool i2s_started;
+	bool rx_started;
+	bool usb_data_received;
+	/* Counter used to determine when to start I2S and then when to start
+	 * sending RX packets to host. Overflows are not a problem because this
+	 * variable is not necessary after both I2S and RX is started.
+	 */
+	uint8_t i2s_counter;
```

Comment:
There are 5 `bool` variables before this `uint8_t`, so moving this `uint8_t` won't really save anything.

On nRF5340DK the `sizeof(struct usb_i2s_ctx)` is 24 bytes. When `i2s_counter` is moved to the end of structure the size remains unchanged at 24 bytes. Moving all pointers to the beginning also makes the size remain intact.

Because there are 9 single-byte members (5 `bool` and 4 `uint8_t`) separated into two "blocks", shuffling them around won't give any structure size reduction regardless if the pointer size is 4 or 8 bytes.

Reply:
Indeed, bool, "the default is CHAR_TYPE_SIZE"